Feature/scrum36 api manage schedule

MarianRoz opened the pull request.

MarianRoz committed:
Added global exeption handling

MarianRoz committed:
Added logger for exeption

MarianRoz committed:
Improved methods "GetAll" and "Update"

bohdansydor commented on ScheduleLesson/Exception/CustomException.cs:
```diff
@@ -0,0 +1,52 @@
+﻿namespace ScheduleLesson
+{
+    public class CustomException : Exception
```
Lets rename to StatusCodeException

bohdansydor replied:
HttpException or smth like that

bohdansydor commented on ScheduleLesson/Exception/CustomException.cs:
```diff
@@ -0,0 +1,52 @@
+            StatusCode = statusCode;
+        }
+    }
+    public class ExceptionHandlingMiddleware
```
Lets put it in separete file

bohdansydor commented on ScheduleLesson/Exception/CustomException.cs:
```diff
@@ -0,0 +1,52 @@
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError($"An unexpected error occurred: {ex.Message}");
```
Where are the logs saved?

bohdansydor replied:
The goal is to store these logs so they can be viewed, used, and analyzed later.

MarianRoz replied:
When an exception occurs, the schedule is not saved anywhere. I don't understand why they should be saved when an exception occurs?

bohdansydor commented on ScheduleLesson/Services/ScheduleService.cs:
```diff
@@ -39,10 +35,9 @@ public async Task<Schedule> UpdateSchedule(Schedule updateSchedule)
                 await _context.SaveChangesAsync();
```
Let's update the method like in "the most important video" :)

bohdansydor replied:
Let's remove try-catch

bohdansydor commented on ScheduleLesson/Program.cs:
```diff
@@ -38,6 +38,7 @@ public static void Main(string[] args)
 
 
             app.MapControllers();
+            app.UseMiddleware<ExceptionHandlingMiddleware>();
```
This is not the ideal place to add this ExceptionHandlingMiddleware. 
.NET 8 pipeline process

MarianRoz committed:
Rename to StatusCodeException

MarianRoz committed:
Posted it in separate file. Refactored cod

MarianRoz committed:
Offset UseMiddleware.
Serilog settings

MarianRoz committed:
Using Logger.
Refactored method GetSchedule and UpdateSchedule

MarianRoz committed:
Refactored UpdateSchedule

bohdansydor commented on ScheduleLesson/Exception/StatusCodeException.cs:
```diff
@@ -0,0 +1,13 @@
+        {
+            StatusCode = statusCode;
+        }
+
```
Lets remove empty lines

bohdansydor commented on ScheduleLesson/Exception/ExceptionHandlingMiddleware.cs:
```diff
@@ -0,0 +1,47 @@
+            catch (Exception ex)
+            {
+                _logger.LogError("Внутрішня помилка сервера");
+                await HandleExceptionAsync(context, 500, ex.Message);
```
Let's use const for 500 status code. Try to avoid 'magic' numbers/strings

bohdansydor commented on ScheduleLesson/Exception/ExceptionHandlingMiddleware.cs:
```diff
@@ -0,0 +1,47 @@
+
+        }
+
+        private static Task HandleExceptionAsync(HttpContext context, int statusCode, string message)
```
Lets clean up this method

bohdansydor commented on ScheduleLesson/Services/ScheduleService.cs:
```diff
@@ -6,45 +6,47 @@ namespace ScheduleLesson.Services
             return await GetSchedule(schedule.Id);
         }
-
         public async Task<Schedule> UpdateSchedule(Schedule updateSchedule)
```
Let's update the method like in "the most important video" :)
One more time)

MarianRoz committed:
Cleared up this method

MarianRoz committed:
Cleared these methods.
Used consts for 500 status code.
Refactored method UpdateSchedule in file ScheduleService

MarianRoz commented:
@bohdansydor  Can I merge branches?

bohdansydor commented on ScheduleLesson/Services/ScheduleService.cs:
```diff
@@ -31,20 +33,29 @@ public async Task<Schedule> AddSchedule(Schedule schedule)
             catch (DbUpdateConcurrencyException ex)
             {
-                throw new StatusCodeException($"Користувач з ID: {updateSchedule.Id} не знайдено", 404);
+                throw new StatusCodeException($"Користувач з ID: {updateSchedule.Id} не знайдено", StatusCodeInternalServerError);
```
Wrong error

bohdansydor commented on ScheduleLesson/Exception/ExceptionHandlingMiddleware.cs:
```diff
@@ -4,12 +4,13 @@ public class ExceptionHandlingMiddleware
     {
         private readonly RequestDelegate _next;
         private readonly ILogger<ExceptionHandlingMiddleware> _logger;
+        private const int StatusCodeInternalServerError = 500;
```
You create 2 the same consts in different files.  
This violates the DRY principle (don't repeat yourself
Find and use ready-made constants from some libs or made public consts in the separate file

bohdansydor commented:
> @bohdansydor Can I merge branches?

when you performed the db logging functionality

MarianRoz committed:
Variable code:
1. Used a ready-made constant from some library
2. Changed the error code

MarianRoz committed:
1. Changed the error code
2. Updated the method "UpdateSchedule"
3. Update appsetting.json
4. Connecting Serilog

MarianRoz committed:
deleted clientapp

MarianRoz committed:
Added System.Net and using it.

MarianRoz commented:
Can I merge the branches now?

MarianRoz commented:
@bohdansydor Can I merge the branches now?

bohdansydor commented on ScheduleLesson/Exception/ExceptionHandlingMiddleware.cs:
```diff
@@ -0,0 +1,39 @@
+        {
+            context.Response.ContentType = "application/json";
+            context.Response.StatusCode = statusCode;
+            var response = new { Message = message };
```
Looks like the variable response is not used

MarianRoz replied:
I suspect it needs to be removed.

bohdansydor commented on ScheduleLesson/Exception/ExceptionHandlingMiddleware.cs:
```diff
@@ -25,7 +25,7 @@ public async Task InvokeAsync(HttpContext context)
             {
                 _logger.LogError("Внутрішня помилка сервера");
-                await HandleExceptionAsync(context, StatusCodeInternalServerError.StatusCode, ex.Message);
+                await HandleExceptionAsync(context, (int)HttpStatusCode.NotFound, ex.Message);
```
Wrong error.
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status

MarianRoz replied:
500?

MarianRoz committed:
Edited error. Deleted variable response.
